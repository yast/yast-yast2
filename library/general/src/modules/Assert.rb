# encoding: utf-8

# ***************************************************************************
#
# Copyright (c) 2002 - 2012 Novell, Inc.
# All Rights Reserved.
#
# This program is free software; you can redistribute it and/or
# modify it under the terms of version 2 of the GNU General Public License as
# published by the Free Software Foundation.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.   See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with this program; if not, contact Novell, Inc.
#
# To contact Novell about this file by physical or electronic mail,
# you may find current contact information at www.novell.com
#
# ***************************************************************************
require "yast"

module Yast
  # Provides assertions for old yast testsuite
  # @deprecated use rspec tests instead
  class AssertClass < Module
    # @param [Object] expected expected value of test
    # @param [Object] actual   actual value of test
    # @param [String] fail_message will be logged if test fails
    # @return whether test succeeds
    def EqualMsg(expected, actual, fail_message)
      expected = deep_copy(expected)
      actual = deep_copy(actual)
      return true if expected == actual

      Builtins.y2error("%1", fail_message)
      false
    end

    # @param [Object] expected expected value of test
    # @param [Object] actual   actual value of test
    # @return whether test succeeds
    def Equal(expected, actual)
      expected = deep_copy(expected)
      actual = deep_copy(actual)
      fail_message = Builtins.sformat(
        "assertion failure, expected '%1', got '%2'",
        expected,
        actual
      )
      EqualMsg(expected, actual, fail_message)
    end

    publish function: :EqualMsg, type: "boolean (any, any, string)"
    publish function: :Equal, type: "boolean (any, any)"
  end

  Assert = AssertClass.new
end
