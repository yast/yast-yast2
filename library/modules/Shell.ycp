/**
 * File:	Shell.ycp
 * Summary:	interface for running shell commands and processing its output.
 * Authors:	Michal Filka <mfilka@suse.cz>
 */
{
module "Shell";

boolean IsEmptyString( string str)
{
    return ( str == nil) || isempty( str);
}

/**
 * @param Shell command to run
 * @return Hash in form $[ "exit": <command-exit-code>, "output": [ <1st line>, <2nd line>, ... ], "stderr": <error output> ] 
 */
global map< string, any> RunAndRead( string command)
{
    map< string, any> ret = $[ "exit": false, "output": [] ];
    map result = ( map) SCR::Execute( .target.bash_output, command);
    string output = result[ "stdout"]:"";

    ret[ "exit"] = result[ "exit"]:nil;

    if( regexpmatch( output, ".*\n$"))
        output = substring( output, 0, size( output) -1 );

    ret[ "output"] = splitstring( output, "\n");
    ret[ "stderr"] = result[ "stderr"]:nil;

    if( ( ret[ "exit"]:nil != 0) || (IsEmptyString( (string) result[ "stderr"]:nil) == false) )
    {
        y2error( "RunAndRead <%1>: Command execution failed. Exit code: %3.\n%2", command, ret[ "exit"]:nil, ret[ "stderr"]:"");
    }

    return ret;
}

/**
 * @param Shell command to run
 * @return whether command execution succeeds
 */
global integer Run( string command)
{
    integer ret = (integer) SCR::Execute( .target.bash, command);

    if( ret != 0)
    {
        y2error( "Run <%1>: Command execution failed. Exit code: %2", command, ret);
    }

    return ret;
}

} /* EOF */
