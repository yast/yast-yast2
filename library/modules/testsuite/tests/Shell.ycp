{
    import "Assert";
    import "Shell";

    Assert::Equal( true, Shell::Run( "echo") );
    Assert::Equal( false, Shell::Run( "false"));

    map< string, any> echo_expected = $[
        "exit": true,
        "output": [ "Hello world!"]
    ];
    map< string, any> echo_multiline_expected = $[
        "exit": true,
        "output": [ "Hello", "world!"]
    ];

    Assert::Equal( echo_expected, Shell::RunAndRead( "echo \"Hello world!\"") );
    Assert::Equal( echo_multiline_expected, Shell::RunAndRead( "echo \"Hello\nworld!\"") );
}
