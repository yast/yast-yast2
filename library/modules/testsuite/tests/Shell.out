Log	Run <false>: Command execution failed.
